### PR TITLE
component: MdInput

### PR DESCRIPTION
**Pull request by hansl**

### CODE REVIEW

File: src/core/annotations/field-value.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ *   <component myField>
+ * or:
+ *   <component myField="">
+ */
+function booleanFieldValueFactory() {
+  return function booleanFieldValueMetadata(target: any, key: string): void {
+    const defaultValue = target[key];
+
+    // Use a fallback if Symbol isn't available.
+    const localKey = isPresent(Symbol) ? Symbol(key) : `@@$_${key}`;
```
Comment (mhevery) [inline src/core/annotations/field-value.ts]: adding special characters such as `@@` makes the object into slow object. Better to use `__private_symbol_` or something like that.

Comment (hansl) [inline src/core/annotations/field-value.ts]: Done. Firefox used to support `@@` as a prefix for symbols, thus my usage of this.

File: src/components/input/input.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  selector: 'md-input',
+  templateUrl: 'components/input/input.html',
+  styleUrls: ['components/input/input.css'],
+  providers: [MdAria, MD_INPUT_CONTROL_VALUE_ACCESSOR],
+  directives: [MdAriaTarget],
+})
+export class MdInput implements ControlValueAccessor {
+  private _empty: boolean = true;
+  private _focused: boolean = false;
+  private _onTouched: () => void = () => {};
```
Comment (ThomasBurleson) [inline src/components/input/input.ts]: Does ng2 provide a noop function like ng1 ?

File: src/components/input/input.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  private _onChangeFn: (_: any) => void = (_: any) => {};
```
Comment (ThomasBurleson) [inline src/components/input/input.ts]: ! WAT ?

File: src/core/utilities/aria-directives.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+import {BooleanFieldValue} from '../annotations/field-value';
+
+
+/**
+ * aria directive, handling aria parts of the standard.
+ */
+@Directive({
+  selector: '[aria-label], [aria-labelledby], [aria-disabled], [aria-required], [aria-invalid]'
+})
+export class MdAria implements AttributeTargetDirectiveInterface {
```
Comment (ThomasBurleson) [inline src/core/utilities/aria-directives.ts]: Does this consider (or is it based upon) the functionality of the ng1 aria service ?

File: src/components/input/input.ts (hunk 3 above)
Comment (hansl) [inline src/components/input/input.ts]: What?

File: src/core/utilities/aria-directives.ts (hunk 4 above)
Comment (hansl) [inline src/core/utilities/aria-directives.ts]: No. This is not a service.

File: src/components/input/input.ts (hunk 3 above)
Comment (ThomasBurleson) [inline src/components/input/input.ts]: again... this is a noop... Perhaps a `noop` Symbol can be created and easily used ?

Comment (m0t0r) [inline src/components/input/input.ts]: very sorry to interfere but angular2 has internal [noop function] (https://github.com/angular/angular/blob/master/modules/angular2/src/facade/lang.ts#L157)

File: src/components/input/input.ts (hunk 2 above)
Comment (hansl) [inline src/components/input/input.ts]: Done.

File: src/components/input/input.ts (hunk 3 above)
Comment (hansl) [inline src/components/input/input.ts]: Okay so after looking deeper, it _is_ invalid Dart to assign a `function(){}` to one that expects an argument. `ts2dart` though emits `dynamic` typing in this case, which remove type checking. I've updated the code to use `noop`, but note this might become invalid in the future.

File: src/components/input/README.md [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# mdInput
+
+Inputs are the basic input component of Material 2. The spec can be found [here](https://www.google.com/design/spec/components/text-fields.html).
+
+
+
+## Type
+
+At the time of writing this README, the `[type]` attribute is copied to the actual `<input>` element in the `<md-input>`. Therefore, the valid types and behavior of those are the ones supported by your browser.
+
```
Comment (zoechi) [inline src/components/input/README.md]: except `file`?

File: src/components/input/input.html [hunk 6]
```diff
@@ -0,0 +1,29 @@
+<div class="md-wrapper" (click)="input.focus()">
```
Comment (jelbourn) [inline src/components/input/input.html]: `md-wrapper` doesn't really tell you anything

File: src/components/input/input.html [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  <div class="md-flex">
+    <prefix><ng-content select="[md-prefix]"></ng-content></prefix>
```
Comment (jelbourn) [inline src/components/input/input.html]: Prefix is not an HTML element; use a css class.
(same for suffix)

File: src/components/input/input.html [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+    <label [attr.for]="'md-input-' + _inputId"
```
Comment (jelbourn) [inline src/components/input/input.html]: Store the entire ID on the class instead of doing the string concat twice in the template?

File: src/components/input/input.html [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+           [class.md-empty]="empty" [class.md-focused]="focused"
+           [class.md-float]="floatingPlaceholder" [class.md-accent]="dividerColor == 'accent'">
+      {{placeholder}}
+      <span *ngIf="required">*</span>
+    </label>
+
+    <input #input aria-target
+           [class.md-end]="align == 'end'"
+           [attr.id]="'md-input-' + _inputId"
+           [disabled]="disabled" [required]="required"
```
Comment (jelbourn) [inline src/components/input/input.html]: One attr per line for scanability?

File: src/components/input/input.html [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+           [attr.maxlength]="maxLength" [attr.type]="type"
```
Comment (jelbourn) [inline src/components/input/input.html]: `maxLength` is a real property on HTMLInputElement, so you don't need the `attr.`

File: src/components/input/input.html [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+
+    <suffix><ng-content select="[md-suffix]"></ng-content></suffix>
+
+  </div>
+
+  <div class="md-underline" [class.md-disabled]="disabled">
+    <span [class.md-focused]="focused" [class.md-accent]="dividerColor == 'accent'"></span>
+  </div>
+  <div *ngIf="hintLabel != ''" class="md-hint">{{hintLabel}}</div>
+  <div *ngIf="_characterCounterLabel() != ''" class="md-hint md-right">{{_characterCounterLabel()}}</div>
```
Comment (jelbourn) [inline src/components/input/input.html]: Private expressions in templates won't work for dart.

File: src/components/input/input.scss [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+    margin: 16px 0;
+  }
+  .md-flex {
+    display: inline-table;
+    flex-flow: column;
+    vertical-align: bottom;
+    width: 100%;
+  }
+  .md-flex > * {
+    display: table-cell;
```
Comment (jelbourn) [inline src/components/input/input.scss]: wat

File: src/components/input/input.scss [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  input {
```
Comment (jelbourn) [inline src/components/input/input.scss]: Can we do these with classes instead of styling the elements directly?

File: src/components/input/input.scss [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+    transition: transform $swift-ease-out-duration $swift-ease-out-timing-function,
+                scale $swift-ease-out-duration $swift-ease-out-timing-function,
+                color $swift-ease-out-duration $swift-ease-out-timing-function;
+
+    &.md-empty {
+      visibility: visible;
+      cursor: text;
+    }
+
+    &.md-float:not(.md-empty), &.md-float.md-focused {
```
Comment (jelbourn) [inline src/components/input/input.scss]: Add comment to explain what this state means

File: src/components/input/input.scss [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+$md-input-underline-disabled-color: md-color($md-foreground, hint-text);
+$md-input-underline-focused-color: md-color($md-primary);
+
+
+:host {
+  display: inline-block;
+  position: relative;
+  font-family: $md-font-family;
+
+  .md-wrapper {
```
Comment (jelbourn) [inline src/components/input/input.scss]: Could you add a brief description of what each class is responsible?

File: src/components/input/input.scss [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+  .md-underline {
+    position: absolute;
+    height: 1px;
+    width: 100%;
+    margin-top: 4px;
+    border-top: 1px solid $md-input-underline-color;
+
+    &.md-disabled {
+      border-top: 0;
+      background-image: linear-gradient(to right, rgba(0,0,0,0.26) 0%, rgba(0,0,0,0.26) 33%, transparent 0%);
```
Comment (jelbourn) [inline src/components/input/input.scss]: All of these background definitions seems like they could benefit from some explanation as to what's happening

File: src/components/input/input.scss [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+      background-position: 0;
+      background-size: 4px 1px;
+      background-repeat: repeat-x;
+    }
+
+    > span {
```
Comment (jelbourn) [inline src/components/input/input.scss]: Definitely a class here; encountering just `> span` is very hard to map back to what's actually being styled in the template.

File: src/components/input/input.ts [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+
+const MD_INPUT_CONTROL_VALUE_ACCESSOR = CONST_EXPR(new Provider(
+    NG_VALUE_ACCESSOR, {
+      useExisting: forwardRef(() => MdInput),
+      multi: true
+    }));
+
+
+let uniqueId = 0;
```
Comment (jelbourn) [inline src/components/input/input.ts]: `nextUniqueId` ?

File: src/components/input/input.html (hunk 7 above)
Comment (hansl) [inline src/components/input/input.html]: Done.

File: src/components/input/input.html (hunk 6 above)
Comment (hansl) [inline src/components/input/input.html]: done.

File: src/components/input/input.ts [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+  @Input() hintLabel: string = '';
+  @Input() maxLength: number = -1;
+  @Input() placeholder: string;
+  @Input() type: string = 'text';
+
+  get value(): any { return this._value; };
+  @Input() set value(v: any) {
+    if (this._value !== v) {
+      this._value = v;
+      this._empty = v === undefined || v === null || (''+v == '');
```
Comment (jelbourn) [inline src/components/input/input.ts]: You should be able to do
```ts
v == null || v == '';
```

File: src/components/input/input.ts [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+export class MdInput implements ControlValueAccessor {
+  private _empty: boolean = true;
+  private _focused: boolean = false;
+  private _onTouched: () => void = noop;
+  private _onChangeFn: (_: any) => void = noop;
+  private _value: any = '';
+  private _inputId: number = uniqueId++;
+
+  get focused() { return this._focused; }
+  get empty() { return this._empty; }
```
Comment (jelbourn) [inline src/components/input/input.ts]: If these don't have setters you don't need getters, just normal properties.

File: src/components/input/input.ts [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+  providers: [MdAria, MD_INPUT_CONTROL_VALUE_ACCESSOR],
+  directives: [MdAriaTarget],
+})
+export class MdInput implements ControlValueAccessor {
+  private _empty: boolean = true;
+  private _focused: boolean = false;
+  private _onTouched: () => void = noop;
+  private _onChangeFn: (_: any) => void = noop;
+  private _value: any = '';
+  private _inputId: number = uniqueId++;
```
Comment (jelbourn) [inline src/components/input/input.ts]: `md-input-${nextUniqueId++}`

File: src/components/input/input.ts [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+@Component({
+  selector: 'md-input',
+  templateUrl: 'components/input/input.html',
+  styleUrls: ['components/input/input.css'],
+  providers: [MdAria, MD_INPUT_CONTROL_VALUE_ACCESSOR],
+  directives: [MdAriaTarget],
+})
+export class MdInput implements ControlValueAccessor {
+  private _empty: boolean = true;
+  private _focused: boolean = false;
```
Comment (jelbourn) [inline src/components/input/input.ts]: `empty` and `focused` to `isEmpty` and `isFocused`?

Comment (jelbourn) [inline src/components/input/input.ts]: Actually, why isn't `isEmpty()` a method? I don't see why you need a property for it since it can be computed from `value`.

File: src/components/input/input.ts [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+  get value(): any { return this._value; };
+  @Input() set value(v: any) {
+    if (this._value !== v) {
+      this._value = v;
+      this._empty = v === undefined || v === null || (''+v == '');
+      this._onChangeFn(v);
+    }
+  }
+
+  @Output('change') change = new EventEmitter(false);
```
Comment (jelbourn) [inline src/components/input/input.ts]: Doesn't `EventEmitter` need a generic?

File: src/components/input/input.ts [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+  @HostBinding('attr.align') private get _align(): any { return null; }
```
Comment (jelbourn) [inline src/components/input/input.ts]: Personally I'm not a fan of singe-lining getters, which I find makes the code more difficult to scan. Especially in this case where the single line has the decorator, its parameter, access modifier, name, type, and the return value.

Also this always returns null- TODO?

File: src/components/input/input.ts [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+
+  /** @override */
```
Comment (jelbourn) [inline src/components/input/input.ts]: Instead of the `override` comment, say 
```
Implemented as part of ControlValueAccessor.
```

File: src/components/input/input.ts (hunk 23 above)
Comment (jelbourn) [inline src/components/input/input.ts]: I don't see this event actually being emitted anywhere.

File: src/components/input/input.ts (hunk 24 above)
Comment (jelbourn) [inline src/components/input/input.ts]: Why use an `attr` instead of a class?

File: src/components/input/input.ts [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    this._focused = true;
+  }
+  onBlur() {
+    this._focused = false;
+    this._onTouched();
+  }
+
+  private _characterCounterLabel(): string {
+    if (this.characterCounter) {
+      return String(this.value || '').length + (this.maxLength > 0 ? ' / ' + this.maxLength : '');
```
Comment (jelbourn) [inline src/components/input/input.ts]: `` `${value}`.length ``?

Comment (jelbourn) [inline src/components/input/input.ts]: Does `X / Y` need to be different for different locales?

File: src/components/input/input.html (hunk 8 above)
Comment (hansl) [inline src/components/input/input.html]: Done.

File: src/components/input/input.html (hunk 9 above)
Comment (hansl) [inline src/components/input/input.html]: Done.

File: src/components/input/input.html (hunk 10 above)
Comment (hansl) [inline src/components/input/input.html]: Actually maxLength is touchy. Done the rest.

File: src/components/input/input.html (hunk 11 above)
Comment (hansl) [inline src/components/input/input.html]: Done.

File: src/components/input/input.scss (hunk 12 above)
Comment (hansl) [inline src/components/input/input.scss]: wat

File: src/core/utilities/aria-directives.ts [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+  Directive,
+  ElementRef,
+  Input,
+  Renderer,
+} from 'angular2/core';
+import {AttributeTargetDirectiveInterface, MdAttributeTarget} from './attribute-target';
+import {BooleanFieldValue} from '../annotations/field-value';
+
+
+/**
```
Comment (jelbourn) [inline src/core/utilities/aria-directives.ts]: I'm concerned about the performance implications of instantiating a directive for every single one of these ARIA attributes.

File: src/components/input/input.scss (hunk 13 above)
Comment (hansl) [inline src/components/input/input.scss]: Done.

File: src/core/utilities/aria-directives.ts (hunk 4 above)
Comment (jelbourn) [inline src/core/utilities/aria-directives.ts]: If we keep this (wrt concerns about performance), the name is too nonspecific. It would be something like `MdAriaAttributeForwarder`

File: src/core/utilities/attribute-target.ts [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+  attributeMap: { [name: string]: any };
+}
+
+
+export class MdAttributeTarget<T extends AttributeTargetDirectiveInterface> {
+  public directive: T;
+  public element: ElementRef;
+  public renderer: Renderer;
+
+  ngAfterViewInit() {
```
Comment (jelbourn) [inline src/core/utilities/attribute-target.ts]: I don't believe this will keep the attributes up to date when the bindings change.

File: src/demo-app/input/input-demo.scss [hunk 29]
```diff
@@ -0,0 +1,23 @@
+@import 'default-theme';
+@import 'variables';
+
+
+.material-icons {
```
Comment (jelbourn) [inline src/demo-app/input/input-demo.scss]: Prefix demo styles with `-demo`. This lets people inspect the elements in the demos and immediately nkow which styles are part of the component itself and which are just from the demo.

File: src/tsconfig.json.backup [hunk 30]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (jelbourn) [inline src/tsconfig.json.backup]: ?

File: src/tsconfig.jsonbackup [hunk 31]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (jelbourn) [inline src/tsconfig.jsonbackup]: ?

**Comment by jelbourn** [comment]
@hansl finished my first pass, let me know when you're ready for another look

File: src/components/input/input.scss (hunk 14 above)
Comment (hansl) [inline src/components/input/input.scss]: Done.

File: src/components/input/input.scss (hunk 15 above)
Comment (hansl) [inline src/components/input/input.scss]: Done.

File: src/components/input/input.scss (hunk 16 above)
Comment (hansl) [inline src/components/input/input.scss]: All of them 1? ;) done.

File: src/components/input/input.scss (hunk 17 above)
Comment (hansl) [inline src/components/input/input.scss]: Done.

File: src/components/input/input.ts (hunk 18 above)
Comment (hansl) [inline src/components/input/input.ts]: Done.

File: src/components/input/input.ts (hunk 19 above)
Comment (hansl) [inline src/components/input/input.ts]: But truthiness is EVIL :(

File: src/components/input/input.ts (hunk 20 above)
Comment (hansl) [inline src/components/input/input.ts]: What do you mean by normal properties? Just `empty: string`? I don't want them to be publicly changeable, they're really readonly.

File: src/components/input/input.ts (hunk 23 above)
Comment (hansl) [inline src/components/input/input.ts]: It's not. Removed.

Comment (hansl) [inline src/components/input/input.ts]: To answer your question; no.

File: src/components/input/input.ts (hunk 24 above)
Comment (hansl) [inline src/components/input/input.ts]: Added a comment to explain why this is needed.

File: src/components/input/input.ts (hunk 25 above)
Comment (hansl) [inline src/components/input/input.ts]: Done. Should we do that for every interface?

File: src/components/input/input.ts (hunk 26 above)
Comment (hansl) [inline src/components/input/input.ts]: Fixed, after offline discussion.

File: src/components/input/input.ts (hunk 19 above)
Comment (jelbourn) [inline src/components/input/input.ts]: The tao of `===` is overrated.

File: src/components/input/input.ts (hunk 21 above)
Comment (hansl) [inline src/components/input/input.ts]: done.

File: src/components/input/input.ts (hunk 22 above)
Comment (hansl) [inline src/components/input/input.ts]: Properties are the same but cleaner IMO. Why should everything be a method, particularly when there's no side effect?

Anyway, done.

File: src/core/utilities/aria-directives.ts (hunk 27 above)
Comment (hansl) [inline src/core/utilities/aria-directives.ts]: After offline discussion; removed these classes and inserted the bindings directly into the input component.

File: src/core/utilities/attribute-target.ts (hunk 28 above)
Comment (hansl) [inline src/core/utilities/attribute-target.ts]: Code has been removed.

File: src/tsconfig.json.backup (hunk 30 above)
Comment (hansl) [inline src/tsconfig.json.backup]: Removed.

File: src/tsconfig.jsonbackup (hunk 31 above)
Comment (hansl) [inline src/tsconfig.jsonbackup]: Not sure what happened there.

**Comment by hansl** [comment]
@jelbourn ready for round #2. What's coming for me; adding unittests as we speak, adding README documentation, adding comments to MdInput.

File: src/components/input/input.ts [hunk 32]
```diff
@@ -0,0 +1,227 @@
...
+})
+export class MdInput implements ControlValueAccessor, AfterContentInit, OnChanges {
+  private _focused: boolean = false;
+  private _onTouched: () => void = noop;
+  private _onChangeFn: (_: any) => void = noop;
+  private _value: any = '';
+
+  /**
+   * Aria related inputs.
+   */
```
Comment (jelbourn) [inline src/components/input/input.ts]: Optional: I'm cool with condensing short block comments to 
```
/** Aria related inputs. */
```

File: src/components/input/input.ts [hunk 33]
```diff
@@ -0,0 +1,227 @@
...
+    return this.placeholder != null || this._placeholderChild != null;
+  }
+
+  private _setValue(value: any): boolean {
+    if (this._value !== value) {
+      this._value = value;
+      return true;
+    } else {
+      return false;
+    }
```
Comment (jelbourn) [inline src/components/input/input.ts]: It's odd that a `set` function returns a value. Why not just emit the event from here?
```ts
_setValue(newValue: any): void {
  if (newValue !== value) {
    this._value = newValue();
    emitChange();
  }
}
```

File: src/components/input/input.ts [hunk 34]
```diff
@@ -0,0 +1,227 @@
...
+ */
+@Directive({
+  selector: 'md-hint',
+  host: {
+    '[class.md-right]': 'align == "end"',
+    '[class.md-hint]': 'true'
+  }
+})
+export class MdHint {
+  @Input() @OneOf(['start', 'end']) align: string;
```
Comment (jelbourn) [inline src/components/input/input.ts]: Add comment for `align`

File: src/components/input/input.ts [hunk 35]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+
+/**
+ * <md-input> component.
```
Comment (jelbourn) [inline src/components/input/input.ts]: Expand description to explain what behaviors it exhibits over a native `<input>`

File: src/components/input/input.ts (hunk 32 above)
Comment (jelbourn) [inline src/components/input/input.ts]: Also, though, I think the `Aria related inputs`, `Content directives`, and `Bindings` section comments don't actually tell you anything. The fields themselves are self-explanatory as it is.

File: src/components/input/input.ts [hunk 36]
```diff
@@ -0,0 +1,227 @@
...
+  @Input('aria-labelledby') ariaLabelledBy: string;
+  @Input('aria-disabled') @BooleanFieldValue() ariaDisabled: boolean;
+  @Input('aria-required') @BooleanFieldValue() ariaRequired: boolean;
+  @Input('aria-invalid') @BooleanFieldValue() ariaInvalid: boolean;
+
+  /**
+   * Content directives.
+   */
+  @ContentChild(MdPlaceholder) private _placeholderChild: MdPlaceholder;
+  @ContentChildren(MdHint) private _hintChildren: QueryList<MdHint>;
```
Comment (jelbourn) [inline src/components/input/input.ts]: You have have multiple hints?

File: src/components/input/input.ts [hunk 37]
```diff
@@ -0,0 +1,227 @@
...
+  @Input() set value(v: any) {
+    if (this._setValue(v)) {
+      this._onChangeFn(v);
+    }
+  }
+
+  // This is to remove the `align` property of the `md-input` itself. Otherwise HTML5
+  // might place it as RTL when we don't want to. We still want to use `align` as an
+  // Input though, so we use HostBinding.
+  @HostBinding('attr.align') private get _align(): any { return null; }
```
Comment (jelbourn) [inline src/components/input/input.ts]: Is this necessary, since we're just using any of the (deprecated) native `align` values?

File: src/components/input/input.ts [hunk 38]
```diff
@@ -0,0 +1,227 @@
...
+  selector: 'md-input',
+  templateUrl: 'components/input/input.html',
+  styleUrls: ['components/input/input.css'],
+  providers: [MdAria, MD_INPUT_CONTROL_VALUE_ACCESSOR],
+  directives: [MdAriaTarget],
+})
+export class MdInput implements ControlValueAccessor, AfterContentInit, OnChanges {
+  private _focused: boolean = false;
+  private _onTouched: () => void = noop;
+  private _onChangeFn: (_: any) => void = noop;
```
Comment (jelbourn) [inline src/components/input/input.ts]: Add comment for `_onChangeFn` and _onTouched` like
```
/** Callback registered via registerOnChange (ControlValueAccessor). */
```
Maybe also rename them to `_onChangeCallback` / `_onTouchedCallback` ?

File: src/components/input/input.ts (hunk 33 above)
Comment (hansl) [inline src/components/input/input.ts]: Unfortunately I cannot do that. There are cases where we want to change the `_value` but not trigger the event; if we do that on a call to `writeValue()` for example, it results in an infinite loop.

File: src/components/input/input.ts [hunk 39]
```diff
@@ -0,0 +1,227 @@
...
+  registerOnChange(fn: any) {
+    this._onChangeFn = fn;
+  }
+
+  /** Implemented as part of ControlValueAccessor. */
+  registerOnTouched(fn: any) {
+    this._onTouched = fn;
+  }
+
+  /** Implemented as part of AfterContentInit. */
```
Comment (jelbourn) [inline src/components/input/input.ts]: I think this comment is unnecessary for lifecycle events.

(I think it's only useful when it's not obvious why the class has the method to begin with)

File: src/components/input/input.ts (hunk 36 above)
Comment (hansl) [inline src/components/input/input.ts]: Left aligned and right aligned.

File: src/components/input/input.ts (hunk 37 above)
Comment (hansl) [inline src/components/input/input.ts]: Yes. I need to remove it from the `<md-input>` tag.

File: src/components/input/input.ts [hunk 40]
```diff
@@ -0,0 +1,227 @@
...
+  ngAfterContentInit() {
+    if (this.placeholder != '' && this.placeholder != null && this._placeholderChild != null) {
```
Comment (jelbourn) [inline src/components/input/input.ts]: Need a high-level comment describing what you need to do _in_ `AfterContentInit`. Part of my wants to say that lifecycle events should only ever just call other functions, so you cleanly separate the *when* and the *what*.

File: src/components/input/input.ts [hunk 41]
```diff
@@ -0,0 +1,227 @@
...
+
+  ngOnChanges(changes: {[key: string]: SimpleChange}) {
+    if (!this._hintChildren) {
+      return;
+    }
+
+    // Initialize the hintChildren.
+    let startHint: MdHint = null;
+    let endHint: MdHint = null;
+    iterateListLike(this._hintChildren, (hint: MdHint) => {
```
Comment (jelbourn) [inline src/components/input/input.ts]: `QueryList` has a `forEach` method.

File: src/components/input/input.ts (hunk 38 above)
Comment (hansl) [inline src/components/input/input.ts]: Done.

File: src/components/input/input.ts (hunk 39 above)
Comment (hansl) [inline src/components/input/input.ts]: Honestly I don't understand why the change of API conventions; `writeValue` should really be `ngWriteValue`, etc... while `AfterContentInit` should be required to be able to call the `ngAfterContentInit` function. :shrug:

File: src/core/annotations/field-value.ts [hunk 42]
```diff
@@ -0,0 +1,38 @@
+import {isPresent} from 'angular2/src/facade/lang';
+
+
+declare var Symbol: any;
```
Comment (jelbourn) [inline src/core/annotations/field-value.ts]: Is this legal? TS already supports ES6 `Symbol` for built-in symbols, so wouldn't there be a conflict? Why use Symbol at all? I'm not sure I see the benefit.

File: src/demo-app/input/input-demo.scss [hunk 43]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+.demo-transform {
+  transition: color $swift-ease-out-duration $swift-ease-out-timing-function;
+}
+.demo-primary {
+  color: md-color($md-primary);
+}
+
+md-card {
```
Comment (jelbourn) [inline src/demo-app/input/input-demo.scss]: moar `.demo-` prefix

File: src/components/input/input.ts (hunk 41 above)
Comment (hansl) [inline src/components/input/input.ts]: mmmh. Apparently it's the `map` function that doesn't work. cool.

File: src/components/input/input.ts (hunk 34 above)
Comment (hansl) [inline src/components/input/input.ts]: done

File: src/components/input/input.scss [hunk 44]
```diff
@@ -0,0 +1,182 @@
...
+
+  // Global wrapper. We need to apply margin to the element for spacing, but
+  // cannot apply it to the host element directly.
+  .md-input-wrapper {
+    margin: 16px 0;
+  }
+
+  // We use a table layout to baseline align the prefix and suffix classes.
+  // The underline is outside of it so it can cover all of the elements under
+  // this table.
```
Comment (jelbourn) [inline src/components/input/input.scss]: Can you talk about why `display: inline-table` is more suitable for this than other approaches (especially flex)?

**Comment by jelbourn** [comment]
Done w/ second pass

File: src/core/annotations/field-value.ts (hunk 42 above)
Comment (hansl) [inline src/core/annotations/field-value.ts]: No. We simply declare the ambient variable. Otherwise it's not defined and TypeScript will scream. Apparently `es6-shim` doesn't have that.

File: src/components/input/input.ts (hunk 40 above)
Comment (hansl) [inline src/components/input/input.ts]: Done. Moved all this.

File: src/demo-app/input/input-demo.scss (hunk 43 above)
Comment (hansl) [inline src/demo-app/input/input-demo.scss]: done.

File: src/components/input/input.scss (hunk 44 above)
Comment (hansl) [inline src/components/input/input.scss]: Took a shot, but my brain is a jello today.

**Comment by hansl** [comment]
PTAL.

File: src/components/input/input.html [hunk 45]
```diff
@@ -0,0 +1,47 @@
...
+  </div>
+
+  <div class="md-input-underline"
+       [class.md-disabled]="disabled">
+    <span class="md-input-ripple"
+          [class.md-focused]="focused"
+          [class.md-accent]="dividerColor == 'accent'"></span>
+  </div>
+
+  <div *ngIf="hintLabel != ''" class="md-hint">{{hintLabel}}</div>
```
Comment (jelbourn) [inline src/components/input/input.html]: No actions now, but my gut is worried about each input having ~25 bindings. Something to explore later.

File: src/components/input/input.scss [hunk 46]
```diff
@@ -0,0 +1,184 @@
...
+    outline: none;
+    padding: 0;
+    width: 100%;
+
+    &.md-end {
+      text-align: right;
+    }
+  }
+
+  // The placeholder label. This is invisible unless it is. The logic to show it is
```
Comment (jelbourn) [inline src/components/input/input.scss]: `"This is invisible unless it is.`"

Deep.

File: src/components/input/input.spec.ts [hunk 47]
```diff
@@ -0,0 +1,180 @@
...
+
+    beforeEach(inject([TestComponentBuilder], function (tcb: TestComponentBuilder) {
+      builder = tcb;
+    }));
+
+    it('creates a native <input> element', injectAsync([], () => {
+      return builder.createAsync(MdInputBaseTestController)
+        .then((fixture) => {
+          fixture.detectChanges();
+          expect(fixture.debugElement.query(By.css('input'))).not.toBeNull();
```
Comment (jelbourn) [inline src/components/input/input.spec.ts]: Would this pass?
```
expect(undefined).not.toBeNull()
```

File: src/components/input/input.spec.ts [hunk 48]
```diff
@@ -0,0 +1,180 @@
...
+          expect(el.getAttribute('aria-label')).toEqual('label');
+          instance.ariaLabel = 'label 2';
+          fixture.detectChanges();
+          expect(el.getAttribute('aria-label')).toEqual('label 2');
+
+          expect(el.getAttribute('aria-disabled')).not.toBeNull();
+        });
+    }));
+
+    it('validates there\'s only one hint label per side', injectAsync([], () => {
```
Comment (jelbourn) [inline src/components/input/input.spec.ts]: Use backticks or double-quotes to avoid escaping?

File: src/components/input/input.ts (hunk 37 above)
Comment (jelbourn) [inline src/components/input/input.ts]: We may want to consider a different attribute name across the library if we're regularly going to have to set up an extra binding just to address this.

File: src/components/input/input.ts (hunk 33 above)
Comment (jelbourn) [inline src/components/input/input.ts]: Can you add that explanation to a comment here?

File: src/components/input/input.spec.ts [hunk 49]
```diff
@@ -0,0 +1,180 @@
...
+          let instance = fixture.componentInstance;
+          fixture.detectChanges();
+          let inputInstance = fixture.debugElement.query(By.directive(MdInput)).componentInstance;
+          expect(inputInstance.characterCount).toEqual(0);
+
+          instance.model = 'hello';
+          fixture.detectChanges();
+          expect(inputInstance.characterCount).toEqual(5);
+        });
+    }));
```
Comment (jelbourn) [inline src/components/input/input.spec.ts]: Should there also be a test for _showing_ the character count?

File: src/components/input/input.spec.ts [hunk 50]
```diff
@@ -0,0 +1,180 @@
...
+
+    it('creates a native <input> element', injectAsync([], () => {
+      return builder.createAsync(MdInputBaseTestController)
+        .then((fixture) => {
+          fixture.detectChanges();
+          expect(fixture.debugElement.query(By.css('input'))).not.toBeNull();
+        });
+    }));
+
+    it('implements ControlValueAccessor', injectAsync([], () => {
```
Comment (jelbourn) [inline src/components/input/input.spec.ts]: Should the test be `it('should support ngModel' ...`, since implementing `ControlValueAccessor` is just an implementation detail?

File: src/components/input/input.spec.ts [hunk 51]
```diff
@@ -0,0 +1,180 @@
...
+    it('validates the type', injectAsync([], () => {
+      return builder.createAsync(MdInputInvalidTypeTestController)
+        .then((fixture: ComponentFixture) => {
+          fakeAsync(() => {
+            expect(() => fixture.detectChanges()).toThrow(MdInputUnsupportedTypeException);
+          });
+        });
+    }));
+  });
+}
```
Comment (jelbourn) [inline src/components/input/input.spec.ts]: Needs tests for various hint and placeholder configurations.

File: src/components/input/input.spec.ts (hunk 49 above)
Comment (hansl) [inline src/components/input/input.spec.ts]: There's no standard for character count, so really it's two tests; one that check the property is what's expected, and the other that checks the hint label is what's expected. The character count is simply a usage of both.

File: src/components/input/input.scss (hunk 46 above)
Comment (hansl) [inline src/components/input/input.scss]: Knew you'd like :)

File: src/components/input/input.html (hunk 45 above)
Comment (hansl) [inline src/components/input/input.html]: Noted. There should be a greater task of performance checks AND visual checking before full release. In beta maybe.

Also, https://www.youtube.com/watch?v=fpaQpyU_QiM

File: src/components/input/input.spec.ts (hunk 48 above)
Comment (hansl) [inline src/components/input/input.spec.ts]: I thought the linting was smart enough to know if you don't need backtick because you don't use `${}`. And double quotes are forbidden by linting too. Will check tomorrow.

File: src/components/input/input.ts (hunk 33 above)
Comment (hansl) [inline src/components/input/input.ts]: Actually after looking at it, now that we don't set a bunch of side variables I can just inline that function.

File: src/components/input/input.ts (hunk 37 above)
Comment (hansl) [inline src/components/input/input.ts]: We may. Let's chat when you come back.

File: src/components/input/input.spec.ts (hunk 47 above)
Comment (hansl) [inline src/components/input/input.spec.ts]: `getAttribute()` returns `null` or  `""`, never `undefined` according to spec.

But yes, that would pass since `toBe` is strict equality. Replaced with `toBeTruthy()`.

File: src/components/input/input.spec.ts (hunk 50 above)
Comment (hansl) [inline src/components/input/input.spec.ts]: Done.

File: src/components/input/input.spec.ts (hunk 51 above)
Comment (hansl) [inline src/components/input/input.spec.ts]: Done. Added a bunch of tests.

**Comment by jelbourn** [comment]
LGTM
